src/main/java/spireMapOverhaul/zones/packattack/PackAttackZone.java

```diff
@@ -0,0 +1,179 @@
+    public boolean canSpawn() {
+        return anniv5 != null && abstractCardPack != null && ((ArrayList<Object>)ReflectionHacks.getPrivateStatic(anniv5, "allPacks")).size() >= 3;
+    }
+
+    @Override
+    public void distributeRooms(Random rng, ArrayList<AbstractRoom> roomList) {
+        //Guarantee at least one event
+        placeRoomRandomly(rng, roomOrDefault(roomList, (room)->room instanceof EventRoom, EventRoom::new));
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public void mapGenDone(ArrayList<ArrayList<MapRoomNode>> map) {
+        ArrayList<Object> allPacks = new ArrayList<>(ReflectionHacks.getPrivateStatic(anniv5, "allPacks"));
+        Collections.shuffle(allPacks, new java.util.Random(AbstractDungeon.mapRng.randomLong()));
+        ArrayList<Object> packs = allPacks.stream().limit(3).collect(Collectors.toCollection(ArrayList::new));
+        this.packNames = new ArrayList<>();
+        this.cards = new ArrayList<>();
+        for (Object pack : packs) {
+            try {
+                this.packNames.add((String)ReflectionHacks.getCachedField(abstractCardPack, "name").get(pack));
+                this.cards.addAll((ArrayList<AbstractCard>)ReflectionHacks.getCachedField(abstractCardPack, "cards").get(pack));
+            } catch (IllegalAccessException e) {
+                e.printStackTrace();
+                throw new RuntimeException("Error determining packs for Pack zone", e);
+            }
+        }
+        this.commonPool = this.cards.stream().filter(c -> c.rarity == AbstractCard.CardRarity.COMMON).collect(Collectors.toCollection(ArrayList::new));
+        this.uncommonPool = this.cards.stream().filter(c -> c.rarity == AbstractCard.CardRarity.UNCOMMON).collect(Collectors.toCollection(ArrayList::new));
+        this.rarePool = this.cards.stream().filter(c -> c.rarity == AbstractCard.CardRarity.RARE).collect(Collectors.toCollection(ArrayList::new));
+        this.updateDescription();
+    }
+
+    @Override
+    public float zoneSpecificEventRate() {
+        return 1;
+    }
+
+    @Override
+    public void atPreBattle() {
+        AbstractDungeon.player.gameHandSize -= 1;
+    }
+
+    @Override
+    public void atTurnStart() {
+        AbstractCard c = this.cards.get(AbstractDungeon.cardRandomRng.random(this.cards.size() - 1)).makeCopy();
+        Wiz.atb(new MakeTempCardInHandAction(c, 1));
+    }
+
+    @Override
+    public void modifyRewardCards(ArrayList<AbstractCard> cards) {
+        this.replaceWithPackCards(cards, AbstractDungeon.cardRng);
+        this.applyStandardUpgradeLogic(cards);
+    }
+
+    @Override
+    public void postCreateShopCards(ArrayList<AbstractCard> coloredCards, ArrayList<AbstractCard> colorlessCards) {
+        this.replaceWithPackCards(coloredCards, AbstractDungeon.merchantRng);
+        for (AbstractCard c : coloredCards) {
+            for (AbstractRelic relic : AbstractDungeon.player.relics) {
+                relic.onPreviewObtainCard(c);
```

Comment:
Why not use applyStandardUpgradeLogic here?

Reply:
This is for shop cards, which don't get the act-dependent update chance.